src/software/multithreading/thread_safe_buffer.tpp

```diff
@@ -0,0 +1,79 @@
+#pragma once
+
+#include "software/logger/logger.h"
+
+template <typename T>
+ThreadSafeBuffer<T>::ThreadSafeBuffer(std::size_t buffer_size)
+    : buffer(buffer_size), destructor_called(false)
+{
+}
+
+template <typename T>
+std::optional<T> ThreadSafeBuffer<T>::popLeastRecentlyAddedValue(Duration max_wait_time)
+{
+    // We hold the returned lock in a variable here so that we hold the lock on the
+    // buffer mutex until the lock is destructed at the end of this function
+    auto buffer_lock = waitForBufferToHaveAValue(max_wait_time);
+
+    std::optional<T> result = std::nullopt;
+    if (!buffer.empty())
+    {
+        result = buffer.front();
+        buffer.pop_front();
+    }
+    return result;
+}
+
+template <typename T>
+std::optional<T> ThreadSafeBuffer<T>::popMostRecentlyAddedValue(Duration max_wait_time)
+{
+    // We hold the returned lock in a variable here so that we hold the lock on the
+    // buffer mutex until the lock is destructed at the end of this function
+    auto buffer_lock = waitForBufferToHaveAValue(max_wait_time);
+
+    std::optional<T> result = std::nullopt;
+    if (!buffer.empty())
+    {
+        result = buffer.back();
+        buffer.pop_back();
+    }
+    return result;
+}
+
+template <typename T>
+void ThreadSafeBuffer<T>::push(const T& value)
+{
+    std::scoped_lock<std::mutex> buffer_lock(buffer_mutex);
+    if (buffer.full())
+    {
+        LOG(DEBUG) << "Pushing to a full ThreadSafeBuffer" << std::endl;
```

Comment:
Can we also log the type of the buffer? That will be way more helpful in identifying where this is happening if we see it fly by in a log.

Reply:
yea that's a good idea. I was also thinking about logging the name of the buffer, but I don't see an easy way to do that

Reply:
I've added the type